graphistry/collapse.py

```diff
@@ -0,0 +1,532 @@
+    :returns new_parent_name
+    """
+    # THIS IS IMPORTANT FUNCTION -- it is where we wrap the parent/child in WRAP
+    # if child in cluster group, we melt it
+    ckey = in_cluster_store_keys(ndf, child)
+
+    if ckey:
+        new_parent_name = melt(ndf, child)
+    else:  # if not, then append child to parent as the start of a new cluster group
+        # might have to escape parent and child if node names are dumb eg, 'this value key'
+        new_parent_name = f"{wrap_key(parent)} {wrap_key(child)}"
+    if VERBOSE:
+        print(f"Renaming parent {parent} with child {child} as {new_parent_name}")
+    return reduce_key(new_parent_name)
+
+
+def collapse_nodes(g, parent, child):
+    """
+        Asserts that parent and child node in ndf should be collapsed into super node.
+        Sets new ndf with COLLAPSE nodes in graphistry instance g
+
+        # this asserts that we SHOULD merge parent and child as super node
+        # outside logic controls when that is the case
+        # for example, it assumes parent is already in cluster keys of COLLAPSE node
+
+    ---------------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param parent: `node` with `attribute` in `column`
+    :param child: `node` with `attribute` in `column`
+    :returns: graphistry instance, new_parent_name of super cluster
+    """
+
+    ndf, edf, src, dst, node = unpack(g)
+
+    new_parent_name = get_new_node_name(ndf, parent, child)
+
+    ndf.loc[ndf[node] == parent, COLLAPSE_NODE] = new_parent_name
+    ndf.loc[ndf[node] == child, COLLAPSE_NODE] = new_parent_name
+
+    g._nodes = ndf
+    return g, new_parent_name
+
+
+def collapse_edges(g, parent, attribute, column, new_parent_name):
+    """
+        Sets new_parent_name to edge list in edf dataframe
+        Finds incluster nodes and adds those as well, but does not add it to seen_dict as that breaks traversal
+
+    --------------------------------------------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param parent: `node` with `attribute` in `column`
+    :param attribute: `attribute` to collapse in `column` over
+    :param column: `column` to collapse over
+    :param new_parent_name: name of collapsed node
+    """
+    # get out of cluster and in cluster nodes from parent node
+    # then feed incluster nodes to new_parent_name
+    outcluster, incluster, tdf = get_edges_in_out_cluster(g, parent, attribute, column)
+    # keep incluster nodes and assign them to new parent
+    ndf, edf, src, dst, node = unpack(g)
+
+    for node in incluster:
+        edf.loc[edf[src] == parent, COLLAPSE_SRC] = new_parent_name
+        edf.loc[edf[dst] == parent, COLLAPSE_DST] = new_parent_name
+
+        edf.loc[edf[dst] == node, COLLAPSE_DST] = new_parent_name
+        edf.loc[edf[src] == node, COLLAPSE_SRC] = new_parent_name
+
+    g._edges = edf
+    return g
```

Comment:
again, this is really a mutation of existing g, not returning a new g, so better to not return anything

Reply:
Fixed